### ISSUE TITLE
BackupPC_backupDelete does not issue an error with incorrect use of `-s` option

### ISSUE THREAD

**Issue by opoplawski** [backuppc/backuppc#519]
I went to try to remove a specific directory from a backup and did:
```
BackupPC_backupDelete -h HOSTNAME -n 742 -s /etc/e2guardian/certs/generatedcerts
```
which is not the correct use of `-s`.  However the command did not complain but proceeded to remove the entire backup:
```
__bpc_pidStart__ 13349
BackupPC_backupDelete: removing #742
__bpc_progress_state__ merge #742 -> #741
BackupPC_backupDelete: Merge into backup 741
__bpc_progress_fileCnt__ 101
__bpc_progress_fileCnt__ 319
__bpc_progress_fileCnt__ 499
...
```

**Comment by GWHAYWOOD** [comment]
Do you think it would be good enough to require that the given share name must not contain the '/' character?

That could be a very easy patch:

```
8<-------------------------------------------------------------------------------------------------------------------------------------------------
--- BackupPC_backupDelete       2020-02-12 11:43:42.611900043 +0000
+++ BackupPC_backupDeleteTest   2025-10-07 17:01:39.208127490 +0100
@@ -190,6 +190,11 @@
 
 if ( $Backups[$idx]{version} < 4 ) {
     if ( defined($ShareName) ) {
+       if ( $ShareName =~ m'/' ) {
+           printf( STDERR "BackupPC_backupDelete: ShareName must not contain '/': [%s]\n", $ShareName );
+           print("__bpc_pidEnd__ $$\n") if ( !$opts{p} );
+           exit(1);
+       }
         foreach my $delPath ( @$DelPaths ) {
             logMsg("BackupPC_backupDelete: removing pre-v4 path #$Backups[$idx]{num}/$ShareName/$delPath\n") if ( $LogLevel >= 1 );
             if ( !$opts{p} ) {
@@ -213,6 +218,11 @@
     }
 } else {
     if ( defined($ShareName) ) {
+       if ( $ShareName =~ m'/' ) {
+           printf( STDERR "BackupPC_backupDelete: ShareName must not contain '/': [%s]\n", $ShareName );
+           print("__bpc_pidEnd__ $$\n") if ( !$opts{p} );
+           exit(1);
+       }
         #
         # Since we are only deleting portions of the backup tree, we
         # need to update reference counts
8<-------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by GWHAYWOOD** [comment]
Here's a possibly better patch.  Not yet much tested.

As in the patch above, there are two places in `BackupPC_backupDelete` where the following line appears:

```
if ( defined($ShareName) ) {
```

After each of these lines insert the following eight lines of code:

```
        my $shareFound = 0;
        foreach my $tempName ( $View->shareList($opts{n}) ) {
            if ( $ShareName eq $tempName ) { $shareFound = 1; last; }
        }
        if ( !$shareFound ) {
            printf( STDERR "Share [%s] not found in backup number %d\n", $ShareName, $opts{n} );
            exit(1);
        }
```
Finally, with all the `use` lines near the top add the line

```
use BackupPC::View;
```

Observations would be welcome.

**Comment by puterboy** [comment]
I like the second patch better... (I also said that in a note to the list)
Why not do the test just once after the line:
```
my $ShareName   = $opts{s};
```

Also, my Perl is getting a bit rusty,  but couldn't you just use Perl grep in the list context, to eliminate the list and the temp variable:
```
unless ( grep { $_ eq $ShareName } $View->shareList($opts{n}) ) {
    printf( STDERR "Share [%s] not found in backup number %d\n", $ShareName, $opts{n} );
    exit(1);
}
```

**Comment by GWHAYWOOD** [comment]
@puterboy thanks for your reply.

Almost everything you say is true, although
I didn't see anything on the list ;
Your Perl doesn't seem to be so rusty :) ;
I'm not sure that your grep will eliminate the list, since it's sub shareList which actually creates it; my loop just iterates over it.  I toyed with the idea of doing something with a hash instead of an array instead, but decided against it.

It's true my coding style will never win a contest; your suggested code change is both elegant and Perlish - mostly why I didn't do it that way. :)

At this stage the issues I'm more concerned with are in this order:
(1) does the patch break anything?
(2) does it fix the problem? and
(3) is there anything else we should be doing with this utility?
these lead me to wonder
(4) how can we get the maximum number of users to test these changes?

Regarding your suggestion to do the test only once and my point (1) above, I want a user to have an easy way to disable the patch either for V3 code or for V4 code in case it breaks something in one version and not the other, which is why there's some duplication.   I also want a user to be able to hack the code with 'print STDERR something' statements if necessary.  Of course, assuming adequate testing and no nasty surprises, the duplication can be eliminated, perhaps together with efficiency/style improvements.

I'm not a big fan of making changes just for reasons of style.  Craig and I recently agreed to remove perltidy from the release scripts.

**Comment by puterboy** [comment]
BTW, on the list I am ***@***.***' -- a long-time user (20+
years) and occasional contributor.

Grep is generally considered faster/more efficient/more elegant than a
loop - though internally, there is of course a loop. And since it is
evaluated in scalar context, it doesn't return a list, I believe.
Again my Perl may be rusty so please verify if you use it...

GWHAYWOOD wrote at about 08:16:36 -0700 on Sunday, October 12, 2025:

**Comment by opoplawski** [comment]
Can we get a proper PR for the fix, please?  Then I will test it.

**Comment by GWHAYWOOD** [comment]
See Draft PR #549.

**Comment by GWHAYWOOD** [comment]
Pushed fix to master.

Sorry, that was typed out of habit.  Not master - backuppc/backuppc branch.

**Comment by opoplawski** [comment]
@GWHAYWOOD so what was the fix?  It would be good if the commit or PR was linked to this issue.  I don't see a recent commit to master and the PR was closed without merging.

**Comment by GWHAYWOOD** [comment]
Sorry, the fix is in 8cfb352afdbeb304c9aaf4be130d8b1234c1df28.

I forked the repo to my private space and made the PR using that; then something else came up which looked urgent (changes to Data::Dumper completely broke BackupPC in master) and I pulled in these edits at the same time.

I've now deleted the private copy of the repo.

**Comment by opoplawski** [comment]
Ah, okay.  That commit is not in `master` but in the `backuppc/backuppc` branch.

**Comment by opoplawski** [comment]
I get:
```
Global symbol "$View" requires explicit package name (did you forget to declare "my $View"?) at /usr/share/BackupPC/bin/BackupPC_backupDelete line 212.
Global symbol "$View" requires explicit package name (did you forget to declare "my $View"?) at /usr/share/BackupPC/bin/BackupPC_backupDelete line 247.
```
So I added it:
```
--- /usr/share/BackupPC/bin/BackupPC_backupDelete.view  2025-10-16 15:55:16.009448206 -0700
+++ /usr/share/BackupPC/bin/BackupPC_backupDelete       2025-10-16 16:03:17.867261659 -0700
@@ -203,6 +203,7 @@
 my $DelPaths    = @ARGV ? [@ARGV] : [undef];
 my $ShareName   = $opts{s};
 my $ShareNameM  = $bpc->fileNameEltMangle($ShareName);
+my $View;

 if ( $Backups[$idx]{version} < 4 ) {
     if ( defined($ShareName) ) {
```
Now I get:
```
__bpc_pidStart__ 34039
Can't call method "shareList" on an undefined value at /usr/share/BackupPC/bin/BackupPC_backupDelete line 248.
```
So then I changed it to:
```
iff -u /usr/share/BackupPC/bin/BackupPC_backupDelete.view /usr/share/BackupPC/bin/BackupPC_backupDelete
--- /usr/share/BackupPC/bin/BackupPC_backupDelete.view  2025-10-16 15:55:16.009448206 -0700
+++ /usr/share/BackupPC/bin/BackupPC_backupDelete       2025-10-16 16:13:09.369613715 -0700
@@ -203,6 +203,7 @@
 my $DelPaths    = @ARGV ? [@ARGV] : [undef];
 my $ShareName   = $opts{s};
 my $ShareNameM  = $bpc->fileNameEltMangle($ShareName);
+my $View = BackupPC::View->new($bpc, $Host, \@Backups);

 if ( $Backups[$idx]{version} < 4 ) {
     if ( defined($ShareName) ) {
```
But it appears that it still goes ahead and removes the full backup:
```
$ /usr/share/BackupPC/bin/BackupPC_backupDelete -h nuc12 -n 5 -s /root__bpc_pidStart__ 34150
BackupPC_backupDelete: removing #5
__bpc_progress_state__ delete #5
BackupPC_backupDelete: No prior backup for merge
__bpc_progress_fileCnt__ 101
...
```
So I would say, no, this isn't fixed yet.

**Comment by GWHAYWOOD** [comment]
> I get:
> 
> ```
> Global symbol "$View" requires explicit package name (did you forget to declare "my $View"?) at /usr/share/BackupPC/bin/BackupPC_backupDelete line 212.
> Global symbol "$View" requires explicit package name (did you forget to declare "my $View"?) at /usr/share/BackupPC/bin/BackupPC_backupDelete line 247.
> ```
> 
> So I added it:

I don't know what's happened here, but that shouldn't have been necessary.  There should be an additional line

`use BackupPC::View;`

with all the other `use` statements near the top of the file (BackupPC_backupDelete).

> So I would say, no, this isn't fixed yet.

I tested it here.  I'm beginning to wonder if I haven't somehow let Github blind-side me.  Wouldn't be the first time.

Unfortunately I can't look at this today, but if you don't figure it out first I'll get to it over the weekend.

**Comment by opoplawski** [comment]
The `use` statement merely allows you to call `BackupPC::View` - it doesn't initialize the `$View` variable.

The other part though it my bad - `/root` actually was a share.  If I change it do a non-existent share, then it properly errors out with:
```
Share [/etc/pki] not found in V4 backup number 5
```

**Comment by GWHAYWOOD** [comment]
Sorry for the mess with this modification.  I **think** it's sorted now.

To recap:

The Data::Dumper issue would have broken BackupPC in the backuppc/backuppc branch so it needed fixing ASAP.

After testing here I somehow grabbed a scratch copy of BackupPC_backupDelete from the wrong directory.

Obviously $View **was** declared and initialized in the working copy but it was not even declared in the out of date copy.

I hope to get this branch tested more - for example AFAIK nobody with a V3 BackupPC has tried it yet - before merging to master.

Apologies again.
